clang/lib/Frontend/PrintPreprocessedOutput.cpp

```diff
@@ -806,6 +837,40 @@ struct UnknownPragmaHandler : public PragmaHandler {
 };
 } // end anonymous namespace
 
+FileID ComputeValidFooterFileID(SourceManager &SM, StringRef Footer) {
+  FileID FooterFileID;
+  llvm::Expected<FileEntryRef> ExpectedFileRef =
+      SM.getFileManager().getFileRef(Footer);
+  if (ExpectedFileRef) {
+    FooterFileID = SM.getOrCreateFileID(ExpectedFileRef.get(),
+                                        SrcMgr::CharacteristicKind::C_User);
+  }
+  assert(FooterFileID.isValid() && "expecting a valid footer FileID");
+  return FooterFileID;
+}
+
+static void PrintIncludeFooter(Preprocessor &PP, SourceLocation Loc,
+                               std::string Footer,
+                               PrintPPOutputPPCallbacks *Callbacks) {
+  SourceManager &SourceMgr = PP.getSourceManager();
+  PresumedLoc UserLoc = SourceMgr.getPresumedLoc(Loc);
+  if (UserLoc.isInvalid())
+    return;
+  FileID FooterFileID = ComputeValidFooterFileID(SourceMgr, Footer);
+  StringRef FooterContentBuffer = SourceMgr.getBufferData(FooterFileID);
+  unsigned NewLine = UserLoc.getLine();
+  // print out the name of the integration footer.
+  Callbacks->WriteFooterInfo(NewLine, Footer);
+  std::string StrippedFooterContentBuffer;
+  SmallVector<StringRef, 8> FooterContentArr;
+  FooterContentBuffer.split(FooterContentArr, '\r');
```

Comment:
Can you explain what this split with carriage-return does?   For my curiosity.  Is this done elsewhere that I could see?

Reply:
split takes a string a returns an array of strings where each element of an array is the content between two separators, in this case \r is the separator.

Reply:
I should have been clearer.   I understand how the function works generally, but was wondering more about where the '\r' gets introduced from.   Do we add it explicitly somewhere else?

Reply:
Sorry I misunderstood your question.
I am not quite sure who adds those. I am assuming that the content of the footer would be some regular C++ code. When I enter artificially in the footer code such that: 
`int foo() {
  int i
    return i++;
}
`
the `FooterContentBuffer` is a `StringRef` where end of lines are represented by `\r\n`.
Do you have by any chance a test case where the footer is actual code, to see concretely in what form the code footer would be?

Reply:
Thanks.  `\r\n` was what I was expecting.   If we split with `\r` won't the part after the split have a `\n`?

For test cases, you could look at: clang/test/CodeGenSYCL/integration-footer.cpp or for a full test, look at: sycl/test-e2e/SpecConstants/2020/non_native/SpecConstBuffer.cpp

Reply:
Yes, it will have.

Reply:
Splitting at \r\n didn't work. So I will leave it as is if you agree.

Reply:
Thanks for checking.   I am fine with leaving it as is.   I was just curious initially.